Some kind of race causing build failure

https://circleci.com/gh/weaveworks/service/2760
This occurred in response to the merging of https://github.com/weaveworks/service/pull/811

The relevant logs:

```
ERRO: 2016/08/30 18:36:23.248705 GET /api/users/logins/mock/attach: Not found
time="2016-08-30T18:36:23Z" level=info msg="GET  (404) 52.942386ms" 
--- FAIL: Test_Account_AttachOauthAccount (19.69s)
    Error Trace:    account_test.go:31
    Error:      Not equal: 200 (expected)
                    != 404 (actual)


    Error Trace:    account_test.go:32
    Error:      Should be true


    Error Trace:    account_test.go:38
    Error:      Not equal: map[string]interface {}{"attach":true, "firstLogin":true} (expected)
                    != map[string]interface {}{"errors":[]interface {}{map[string]interface {}{"message":"Not found"}}} (actual)


    Error Trace:    account_test.go:43
    Error:      No error is expected but got Not found


INFO: 2016/08/30 18:36:23.251135 Running Database Migrations...
--- FAIL: Test_Account_AttachOauthAccount_AlreadyAttachedToAnotherAccount (0.13s)
    Error Trace:    helpers.go:19
            helpers_test.go:30
            account_test.go:58
    Error:      No error is expected but got pq: duplicate key value violates unique constraint "users_lower_email_idx"


INFO: 2016/08/30 18:36:23.377170 Running Database Migrations...
time="2016-08-30T18:36:23Z" level=info msg="GET  (200) 5.37738ms" 
INFO: 2016/08/30 18:36:23.478413 Running Database Migrations...
time="2016-08-30T18:36:23Z" level=info msg="GET  (200) 1.693709ms" 
time="2016-08-30T18:36:23Z" level=info msg="GET  (200) 1.751223ms" 
INFO: 2016/08/30 18:36:23.566824 Running Database Migrations...
time="2016-08-30T18:36:23Z" level=info msg="POST  (204) 2.756878ms" 
INFO: 2016/08/30 18:36:23.636326 Running Database Migrations...
time="2016-08-30T18:36:23Z" level=info msg="POST  (200) 11.970961ms" 
INFO: 2016/08/30 18:36:23.711655 Running Database Migrations...
time="2016-08-30T18:36:23Z" level=info msg="POST  (200) 11.894633ms" 
INFO: 2016/08/30 18:36:23.799524 Running Database Migrations...
ERRO: 2016/08/30 18:36:23.876044 POST /api/users/org/solitary-feather-45/users: invalid character 'g' looking for beginning of value
time="2016-08-30T18:36:23Z" level=info msg="POST  (400) 2.052803ms" 
INFO: 2016/08/30 18:36:23.876376 Running Database Migrations...
ERRO: 2016/08/30 18:36:23.973188 POST /api/users/org/restless-wind-66/users: Email cannot be blank
time="2016-08-30T18:36:23Z" level=info msg="POST  (400) 3.52693ms" 
INFO: 2016/08/30 18:36:23.973474 Running Database Migrations...
time="2016-08-30T18:36:24Z" level=info msg="POST  (200) 16.406943ms" 
INFO: 2016/08/30 18:36:24.108720 Running Database Migrations...
ERRO: 2016/08/30 18:36:24.264364 POST /api/users/org/wispy-nebula-08/users: Forbidden found
time="2016-08-30T18:36:24Z" level=info msg="POST  (403) 2.401448ms" 
INFO: 2016/08/30 18:36:24.266390 Running Database Migrations...
time="2016-08-30T18:36:24Z" level=info msg="POST  (200) 17.442047ms" 
INFO: 2016/08/30 18:36:24.407608 Running Database Migrations...
time="2016-08-30T18:36:24Z" level=info msg="POST  (200) 15.586315ms" 
INFO: 2016/08/30 18:36:24.553225 Running Database Migrations...
time="2016-08-30T18:36:24Z" level=info msg="DELETE  (204) 3.509647ms" 
time="2016-08-30T18:36:24Z" level=info msg="GET  (200) 3.923131ms" 
INFO: 2016/08/30 18:36:24.676027 Running Database Migrations...
time="2016-08-30T18:36:24Z" level=info msg="DELETE  (204) 3.419842ms" 
INFO: 2016/08/30 18:36:24.791743 Running Database Migrations...
ERRO: 2016/08/30 18:36:24.882918 DELETE /api/users/org/aged-snowflake-51/users/6281838661429879825@weave.works: Forbidden found
time="2016-08-30T18:36:24Z" level=info msg="DELETE  (403) 2.303206ms" 
INFO: 2016/08/30 18:36:24.884890 Running Database Migrations...
ERRO: 2016/08/30 18:36:24.993437 DELETE /api/users/org/foobar/users/6842348953158377901@weave.works: Not found
time="2016-08-30T18:36:24Z" level=info msg="DELETE  (404) 2.461383ms" 
INFO: 2016/08/30 18:36:24.993760 Running Database Migrations...
ERRO: 2016/08/30 18:36:25.052793 GET /api/users/login: Email cannot be blank
time="2016-08-30T18:36:25Z" level=info msg="GET  (400) 167.937µs" 
INFO: 2016/08/30 18:36:25.053141 Running Database Migrations...
ERRO: 2016/08/30 18:36:25.119877 GET /api/users/login: Invalid authentication data
time="2016-08-30T18:36:25Z" level=info msg="GET  (401) 1.655328ms" 
INFO: 2016/08/30 18:36:25.120102 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (200) 1.886274ms" 
INFO: 2016/08/30 18:36:25.200863 Running Database Migrations...
ERRO: 2016/08/30 18:36:25.308177 GET /private/api/users/lookup/fooorg: Invalid authentication data
time="2016-08-30T18:36:25Z" level=info msg="GET  (401) 1.694397ms" 
INFO: 2016/08/30 18:36:25.308510 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (200) 1.919213ms" 
INFO: 2016/08/30 18:36:25.399508 Running Database Migrations...
ERRO: 2016/08/30 18:36:25.502751 GET /api/users/lookup: Invalid authentication data
time="2016-08-30T18:36:25Z" level=info msg="GET  (401) 2.821918ms" 
INFO: 2016/08/30 18:36:25.503065 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (200) 2.028918ms" 
INFO: 2016/08/30 18:36:25.603802 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (200) 1.733619ms" 
INFO: 2016/08/30 18:36:25.718260 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (401) 1.876299ms" 
INFO: 2016/08/30 18:36:25.822069 Running Database Migrations...
time="2016-08-30T18:36:25Z" level=info msg="GET  (200) 1.712835ms" 
INFO: 2016/08/30 18:36:25.935554 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="GET  (200) 1.960828ms" 
INFO: 2016/08/30 18:36:26.035734 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="GET  (200) 4.384693ms" 
INFO: 2016/08/30 18:36:26.146472 Running Database Migrations...
ERRO: 2016/08/30 18:36:26.214003 PUT /api/users/org/brilliant-butterfly-03: Forbidden found
time="2016-08-30T18:36:26Z" level=info msg="PUT  (403) 2.133079ms" 
ERRO: 2016/08/30 18:36:26.222538 PUT /api/users/org/not-found-org: Not found
time="2016-08-30T18:36:26Z" level=info msg="PUT  (404) 2.316199ms" 
time="2016-08-30T18:36:26Z" level=info msg="PUT  (204) 2.822955ms" 
INFO: 2016/08/30 18:36:26.227372 Running Database Migrations...
ERRO: 2016/08/30 18:36:26.283559 PUT /api/users/org/snowy-stone-25: ID cannot be changed
time="2016-08-30T18:36:26Z" level=info msg="PUT  (400) 1.81041ms" 
INFO: 2016/08/30 18:36:26.285469 Running Database Migrations...
ERRO: 2016/08/30 18:36:26.348994 PUT /api/users/org/twilight-blossom-87: Name cannot be blank
time="2016-08-30T18:36:26Z" level=info msg="PUT  (400) 1.941314ms" 
INFO: 2016/08/30 18:36:26.350849 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="POST  (201) 4.347743ms" 
INFO: 2016/08/30 18:36:26.422172 Running Database Migrations...
ERRO: 2016/08/30 18:36:26.484656 POST /api/users/org: ID is already taken
time="2016-08-30T18:36:26Z" level=info msg="POST  (400) 2.520599ms" 
ERRO: 2016/08/30 18:36:26.488384 POST /api/users/org: ID cannot be blank
time="2016-08-30T18:36:26Z" level=info msg="POST  (400) 1.884252ms" 
ERRO: 2016/08/30 18:36:26.491892 POST /api/users/org: ID can only contain letters, numbers, hyphen, and underscore
time="2016-08-30T18:36:26Z" level=info msg="POST  (400) 1.839137ms" 
INFO: 2016/08/30 18:36:26.493757 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="GET  (200) 2.409114ms" 
INFO: 2016/08/30 18:36:26.552470 Running Database Migrations...
ERRO: 2016/08/30 18:36:26.616991 GET /api/users/org/brief-resonance-71: Not found
time="2016-08-30T18:36:26Z" level=info msg="GET  (404) 2.07928ms" 
time="2016-08-30T18:36:26Z" level=info msg="GET  (403) 2.081813ms" 
INFO: 2016/08/30 18:36:26.622082 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="POST  (201) 4.51981ms" 
time="2016-08-30T18:36:26Z" level=info msg="POST  (201) 4.472746ms" 
INFO: 2016/08/30 18:36:26.688096 Running Database Migrations...
time="2016-08-30T18:36:26Z" level=info msg="DELETE  (204) 2.491548ms" 
time="2016-08-30T18:36:26Z" level=info msg="DELETE  (403) 2.218696ms" 
time="2016-08-30T18:36:26Z" level=info msg="DELETE  (204) 3.20086ms" 
INFO: 2016/08/30 18:36:26.771616 Running Database Migrations...
INFO: 2016/08/30 18:36:26.846790 Running Database Migrations...
INFO: 2016/08/30 18:36:26.937009 Running Database Migrations...
INFO: 2016/08/30 18:36:27.030271 Running Database Migrations...
INFO: 2016/08/30 18:36:27.115402 Running Database Migrations...
time="2016-08-30T18:36:27Z" level=info msg="POST  (200) 17.803387ms" 
time="2016-08-30T18:36:27Z" level=info msg="GET  (200) 9.714982ms" 
time="2016-08-30T18:36:27Z" level=info msg="POST  (200) 8.212207ms" 
time="2016-08-30T18:36:27Z" level=info msg="GET  (200) 6.505033ms" 
INFO: 2016/08/30 18:36:27.247320 Running Database Migrations...
ERRO: 2016/08/30 18:36:27.310042 POST /api/users/signup: invalid character 'h' in literal true (expecting 'r')
time="2016-08-30T18:36:27Z" level=info msg="POST  (400) 134.245µs" 
INFO: 2016/08/30 18:36:27.310885 Running Database Migrations...
ERRO: 2016/08/30 18:36:27.386874 POST /api/users/signup: Email cannot be blank
time="2016-08-30T18:36:27Z" level=info msg="POST  (400) 109.584µs" 
INFO: 2016/08/30 18:36:27.387621 Running Database Migrations...
time="2016-08-30T18:36:27Z" level=info msg="GET  (200) 11.399705ms" 
INFO: 2016/08/30 18:36:27.471927 Running Database Migrations...
time="2016-08-30T18:36:27Z" level=info msg="GET  (200) 10.083714ms" 
INFO: 2016/08/30 18:36:27.542243 Running Database Migrations...
time="2016-08-30T18:36:27Z" level=info msg="GET  (200) 5.478727ms" 
FAIL
FAIL    github.com/weaveworks/service/users/api 24.111s
ok      github.com/weaveworks/service/users/client  0.028s
?       github.com/weaveworks/service/users/cmd/users   [no test files]
?       github.com/weaveworks/service/users/emailer [no test files]
?       github.com/weaveworks/service/users/externalIDs [no test files]
?       github.com/weaveworks/service/users/login   [no test files]
ok      github.com/weaveworks/service/users/pardot  0.007s
?       github.com/weaveworks/service/users/render  [no test files]
?       github.com/weaveworks/service/users/sessions    [no test files]
time="2016-08-30T18:36:08Z" level=info msg="Running Database Migrations..." 
--- FAIL: Test_Storage_RemoveOtherUsersAccess (14.32s)
    Error Trace:    helpers.go:19
            helpers.go:29
            storage_test.go:15
    Error:      No error is expected but got pq: duplicate key value violates unique constraint "users_lower_email_idx"


time="2016-08-30T18:36:23Z" level=info msg="Running Database Migrations..." 
FAIL
FAIL    github.com/weaveworks/service/users/storage 14.515s
```

most likely only the first error is important, the rest are flow-on effects.

--- a/users/db/db.go
+++ b/users/db/db.go
@@ -94,11 +94,6 @@ type DB interface {
 	Close() error
 }
 
-// Truncater is used in testing, but is not part of the normal db interface
-type Truncater interface {
-	Truncate() error
-}
-
 // MustNew creates a new database from the URI, or panics.
 func MustNew(databaseURI, migrationsDir string) DB {
 	u, err := url.Parse(databaseURI)

--- a/users/db/dbtest/helpers.go
+++ b/users/db/dbtest/helpers.go
@@ -5,29 +5,12 @@ import (
 	"math/rand"
 	"testing"
 
-	"golang.org/x/crypto/bcrypt"
-
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
-	"github.com/weaveworks/service/common/logging"
 	"github.com/weaveworks/service/users"
 	"github.com/weaveworks/service/users/db"
 )
-
-// Setup sets up stuff for testing, creating a new database
-func Setup(t *testing.T) db.DB {
-	require.NoError(t, logging.Setup("debug"))
-	db.PasswordHashingCost = bcrypt.MinCost
-	database := db.MustNew(*databaseURI, *databaseMigrations)
-	require.NoError(t, database.(db.Truncater).Truncate())
-	return database
-}
-
-// Cleanup cleans up after a test
-func Cleanup(t *testing.T, database db.DB) {
-	require.NoError(t, database.Close())
-}
 
 // GetUser makes a randomly named user
 func GetUser(t *testing.T, db db.DB) *users.User {

--- a/users/db/dbtest/integration.go
+++ b/users/db/dbtest/integration.go
@@ -2,9 +2,55 @@
 
 package dbtest
 
-import "flag"
+import (
+	"flag"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/weaveworks/service/common/logging"
+	"github.com/weaveworks/service/users/db"
+	"github.com/weaveworks/service/users/db/postgres"
+)
 
 var (
 	databaseURI        = flag.String("database-uri", "postgres://postgres@users-db.weave.local/users_test?sslmode=disable", "Uri of a test database")
 	databaseMigrations = flag.String("database-migrations", "/migrations", "Path where the database migration files can be found")
+
+	done        chan error
+	errRollback = fmt.Errorf("Rolling back test data")
 )
+
+// Setup sets up stuff for testing, creating a new database
+func Setup(t *testing.T) db.DB {
+	require.NoError(t, logging.Setup("debug"))
+	db.PasswordHashingCost = bcrypt.MinCost
+	// Don't use db.MustNew, here so we can do a transaction around the whole test, to rollback.
+	pg, err := postgres.New(*databaseURI, *databaseMigrations, bcrypt.MinCost)
+	require.NoError(t, err)
+
+	newDB := make(chan db.DB)
+	done = make(chan error)
+	go func() {
+		done <- pg.Transaction(func(tx postgres.DB) error {
+			// Pass out the tx so we can run the test
+			newDB <- tx
+			// Wait for the test to finish
+			return <-done
+		})
+	}()
+	// Get the new database
+	return <-newDB
+}
+
+// Cleanup cleans up after a test
+func Cleanup(t *testing.T, database db.DB) {
+	if done != nil {
+		done <- errRollback
+		require.Equal(t, errRollback, <-done)
+		done = nil
+	}
+	require.NoError(t, database.Close())
+}

--- a/users/db/dbtest/unit.go
+++ b/users/db/dbtest/unit.go
@@ -2,9 +2,31 @@
 
 package dbtest
 
-import "flag"
+import (
+	"flag"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/weaveworks/service/common/logging"
+	"github.com/weaveworks/service/users/db"
+)
 
 var (
 	databaseURI        = flag.String("database-uri", "memory://", "Uri of a test database")
 	databaseMigrations = flag.String("database-migrations", "", "Path where the database migration files can be found")
 )
+
+// Setup sets up stuff for testing, creating a new database
+func Setup(t *testing.T) db.DB {
+	require.NoError(t, logging.Setup("debug"))
+	db.PasswordHashingCost = bcrypt.MinCost
+	database := db.MustNew(*databaseURI, *databaseMigrations)
+	return database
+}
+
+// Cleanup cleans up after a test
+func Cleanup(t *testing.T, database db.DB) {
+	require.NoError(t, database.Close())
+}

--- a/users/db/memory/memory.go
+++ b/users/db/memory/memory.go
@@ -30,19 +30,6 @@ func New(_, _ string, passwordHashingCost int) (*DB, error) {
 	}, nil
 }
 
-// Truncate clears all the data. Should only be used in tests!
-func (d *DB) Truncate() error {
-	*d = DB{
-		users:               make(map[string]*users.User),
-		organizations:       make(map[string]*users.Organization),
-		memberships:         make(map[string][]string),
-		logins:              make(map[string]*login.Login),
-		apiTokens:           make(map[string]*users.APIToken),
-		passwordHashingCost: d.passwordHashingCost,
-	}
-	return nil
-}
-
 // Close finishes using the db. Noop.
 func (d *DB) Close() error {
 	return nil

--- a/users/db/postgres/api_token.go
+++ b/users/db/postgres/api_token.go
@@ -17,7 +17,7 @@ func (d DB) CreateAPIToken(userID, description string) (*users.APIToken, error) 
 		CreatedAt:   d.Now(),
 	}
 
-	err := d.Transaction(func(tx *sql.Tx) error {
+	err := d.Transaction(func(tx DB) error {
 		for exists := t.Token == ""; exists; {
 			if err := t.RegenerateToken(); err != nil {
 				return err

--- a/users/db/postgres/organization.go
+++ b/users/db/postgres/organization.go
@@ -37,15 +37,10 @@ func (d DB) RemoveUserFromOrganization(orgExternalID, email string) error {
 
 // UserIsMemberOf checks if the user is a member of the organization
 func (d DB) UserIsMemberOf(userID, orgExternalID string) (bool, error) {
-	return d.userIsMemberOf(d.DB, userID, orgExternalID)
-}
-
-func (d DB) userIsMemberOf(db squirrel.BaseRunner, userID, orgExternalID string) (bool, error) {
 	rows, err := d.organizationsQuery().
 		Join("memberships on (organizations.id = memberships.organization_id)").
 		Where(squirrel.Eq{"memberships.user_id": userID, "organizations.external_id": orgExternalID}).
 		Where("memberships.deleted_at is null").
-		RunWith(db).
 		Query()
 	if err != nil {
 		return false, err
@@ -117,8 +112,8 @@ func (d DB) ListOrganizationsForUserIDs(userIDs ...string) ([]*users.Organizatio
 	return orgs, err
 }
 
-func (d DB) addUserToOrganization(db execQueryRower, userID, organizationID string) error {
-	_, err := db.Exec(`
+func (d DB) addUserToOrganization(userID, organizationID string) error {
+	_, err := d.Exec(`
 			insert into memberships
 				(user_id, organization_id, created_at)
 				values ($1, $2, $3)`,
@@ -144,10 +139,10 @@ func (d DB) GenerateOrganizationExternalID() (string, error) {
 		externalID string
 		err        error
 	)
-	err = d.Transaction(func(tx *sql.Tx) error {
+	err = d.Transaction(func(tx DB) error {
 		for exists := true; exists; {
 			externalID = externalIDs.Generate()
-			exists, err = d.organizationExists(tx, externalID)
+			exists, err = tx.OrganizationExists(externalID)
 		}
 		return nil
 	})
@@ -165,8 +160,8 @@ func (d DB) CreateOrganization(ownerID, externalID, name string) (*users.Organiz
 		return nil, err
 	}
 
-	err := d.Transaction(func(tx *sql.Tx) error {
-		if exists, err := d.organizationExists(tx, o.ExternalID); err != nil {
+	err := d.Transaction(func(tx DB) error {
+		if exists, err := tx.OrganizationExists(o.ExternalID); err != nil {
 			return err
 		} else if exists {
 			return users.ErrOrgExternalIDIsTaken
@@ -193,7 +188,7 @@ func (d DB) CreateOrganization(ownerID, externalID, name string) (*users.Organiz
 			return err
 		}
 
-		return d.addUserToOrganization(tx, ownerID, o.ID)
+		return tx.addUserToOrganization(ownerID, o.ID)
 	})
 	if err != nil {
 		return nil, err
@@ -206,9 +201,9 @@ func (d DB) CreateOrganization(ownerID, externalID, name string) (*users.Organiz
 func (d DB) FindOrganizationByProbeToken(probeToken string) (*users.Organization, error) {
 	var o *users.Organization
 	var err error
-	err = d.Transaction(func(tx *sql.Tx) error {
-		o, err = d.scanOrganization(
-			d.organizationsQuery().RunWith(tx).Where(squirrel.Eq{"organizations.probe_token": probeToken}).QueryRow(),
+	err = d.Transaction(func(tx DB) error {
+		o, err = tx.scanOrganization(
+			tx.organizationsQuery().Where(squirrel.Eq{"organizations.probe_token": probeToken}).QueryRow(),
 		)
 		if err == nil && o.FirstProbeUpdateAt.IsZero() {
 			o.FirstProbeUpdateAt = d.Now()
@@ -283,12 +278,8 @@ func (d DB) RenameOrganization(externalID, name string) error {
 // OrganizationExists just returns a simple bool checking if an organization
 // exists
 func (d DB) OrganizationExists(externalID string) (bool, error) {
-	return d.organizationExists(d, externalID)
-}
-
-func (d DB) organizationExists(db queryRower, externalID string) (bool, error) {
 	var exists bool
-	err := db.QueryRow(
+	err := d.QueryRow(
 		`select exists(select 1 from organizations where lower(external_id) = lower($1) and deleted_at is null)`,
 		externalID,
 	).Scan(&exists)

--- a/users/db/postgres/postgres.go
+++ b/users/db/postgres/postgres.go
@@ -14,42 +14,38 @@ import (
 
 // DB is a postgres db, for dev and production
 type DB struct {
-	*sql.DB
+	dbProxy
 	squirrel.StatementBuilderType
 	PasswordHashingCost int
 }
 
-type queryRower interface {
-	QueryRow(query string, args ...interface{}) *sql.Row
-}
-
-type execer interface {
+type dbProxy interface {
 	Exec(query string, args ...interface{}) (sql.Result, error)
-}
-
-type execQueryRower interface {
-	execer
-	queryRower
+	Query(query string, args ...interface{}) (*sql.Rows, error)
+	QueryRow(query string, args ...interface{}) *sql.Row
+	Prepare(query string) (*sql.Stmt, error)
 }
 
 // New creates a new postgred DB
-func New(databaseURI, migrationsDir string, passwordHashingCost int) (*DB, error) {
+func New(databaseURI, migrationsDir string, passwordHashingCost int) (DB, error) {
 	if migrationsDir != "" {
 		logrus.Infof("Running Database Migrations...")
 		if errs, ok := migrate.UpSync(databaseURI, migrationsDir); !ok {
 			for _, err := range errs {
 				logrus.Error(err)
 			}
-			return nil, errors.New("Database migrations failed")
+			return DB{}, errors.New("Database migrations failed")
 		}
 	}
 	db, err := sql.Open("postgres", databaseURI)
-	return &DB{
-		DB:                   db,
-		StatementBuilderType: squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar).RunWith(db),
+	return DB{
+		dbProxy:              db,
+		StatementBuilderType: statementBuilder(db),
 		PasswordHashingCost:  passwordHashingCost,
 	}, err
 }
+
+var statementBuilder = squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar).RunWith
 
 // Now gives us the current time for Postgres. Postgres only stores times to
 // the microsecond, so we pre-truncate times so tests will match. We also
@@ -60,12 +56,22 @@ func (d DB) Now() time.Time {
 
 // Transaction runs the given function in a postgres transaction. If fn returns
 // an error the txn will be rolled back.
-func (d DB) Transaction(f func(*sql.Tx) error) error {
-	tx, err := d.Begin()
+func (d DB) Transaction(f func(DB) error) error {
+	if _, ok := d.dbProxy.(*sql.Tx); ok {
+		// Already in a nested transaction
+		return f(d)
+	}
+
+	tx, err := d.dbProxy.(*sql.DB).Begin()
 	if err != nil {
 		return err
 	}
-	if err = f(tx); err != nil {
+	err = f(DB{
+		dbProxy:              tx,
+		StatementBuilderType: statementBuilder(tx),
+		PasswordHashingCost:  d.PasswordHashingCost,
+	})
+	if err != nil {
 		// Rollback error is ignored as we already have one in progress
 		if err2 := tx.Rollback(); err2 != nil {
 			logrus.Warn("transaction rollback: %v (ignored)", err2)
@@ -75,25 +81,12 @@ func (d DB) Transaction(f func(*sql.Tx) error) error {
 	return tx.Commit()
 }
 
-// Truncate clears all the data in pg. Should only be used in tests!
-func (d DB) Truncate() error {
-	return d.Transaction(func(tx *sql.Tx) error {
-		return mustExec(
-			tx,
-			`truncate table traceable;`,
-			`truncate table users;`,
-			`truncate table logins;`,
-			`truncate table organizations;`,
-			`truncate table memberships;`,
-		)
-	})
-}
-
-func mustExec(e squirrel.Execer, queries ...string) error {
-	for _, q := range queries {
-		if _, err := e.Exec(q); err != nil {
-			return err
-		}
+// Close finishes using the db
+func (d DB) Close() error {
+	if db, ok := d.dbProxy.(interface {
+		Close() error
+	}); ok {
+		return db.Close()
 	}
 	return nil
 }

--- a/users/db/postgres/user.go
+++ b/users/db/postgres/user.go
@@ -14,12 +14,8 @@ import (
 
 // CreateUser creates a new user with the given email.
 func (d DB) CreateUser(email string) (*users.User, error) {
-	return d.createUser(d, email)
-}
-
-func (d DB) createUser(q queryRower, email string) (*users.User, error) {
 	u := &users.User{Email: email, CreatedAt: d.Now()}
-	err := q.QueryRow("insert into users (email, approved_at, created_at) values (lower($1), $2, $2) returning id", email, u.CreatedAt).Scan(&u.ID)
+	err := d.QueryRow("insert into users (email, approved_at, created_at) values (lower($1), $2, $2) returning id", email, u.CreatedAt).Scan(&u.ID)
 	switch {
 	case err == sql.ErrNoRows:
 		return nil, users.ErrNotFound
@@ -45,18 +41,17 @@ func (d DB) AddLoginToUser(userID, provider, providerID string, session json.Raw
 		}
 		values["session"] = sessionJSON
 	}
-	return d.Transaction(func(tx *sql.Tx) error {
+	return d.Transaction(func(tx DB) error {
 		// check that this is not already attached somewhere else
-		existing, err := d.findUserByLogin(tx, provider, providerID)
+		existing, err := tx.FindUserByLogin(provider, providerID)
 		switch err {
 		case nil:
 			if existing.ID != userID {
 				return users.AlreadyAttachedError{ID: existing.ID, Email: existing.Email}
 			}
 			// User is already attached to this auth provider, just update the session
-			_, err = d.
+			_, err = tx.
 				Update("logins").
-				RunWith(tx).
 				Where(squirrel.Eq{
 					"user_id":     userID,
 					"provider":    provider,
@@ -68,9 +63,8 @@ func (d DB) AddLoginToUser(userID, provider, providerID string, session json.Raw
 			err = nil
 			// User is not attached to this auth provider, attach them
 			values["created_at"] = now
-			_, err = d.
+			_, err = tx.
 				Insert("logins").
-				RunWith(tx).
 				SetMap(values).
 				Exec()
 		}
@@ -97,32 +91,32 @@ func (d DB) DetachLoginFromUser(userID, provider string) error {
 func (d DB) InviteUser(email, orgExternalID string) (*users.User, bool, error) {
 	var u *users.User
 	userCreated := false
-	err := d.Transaction(func(tx *sql.Tx) error {
-		o, err := d.scanOrganization(
-			d.organizationsQuery().RunWith(tx).Where("lower(organizations.external_id) = lower($1)", orgExternalID).QueryRow(),
+	err := d.Transaction(func(tx DB) error {
+		o, err := tx.scanOrganization(
+			tx.organizationsQuery().Where("lower(organizations.external_id) = lower($1)", orgExternalID).QueryRow(),
 		)
 		if err != nil {
 			return err
 		}
 
-		u, err = d.findUserByEmail(tx, email)
+		u, err = tx.FindUserByEmail(email)
 		if err == users.ErrNotFound {
-			u, err = d.createUser(tx, email)
+			u, err = tx.CreateUser(email)
 			userCreated = true
 		}
 		if err != nil {
 			return err
 		}
 
-		isMember, err := d.userIsMemberOf(tx, u.ID, orgExternalID)
+		isMember, err := tx.UserIsMemberOf(u.ID, orgExternalID)
 		if err != nil || isMember {
 			return err
 		}
-		err = d.addUserToOrganization(tx, u.ID, o.ID)
+		err = tx.addUserToOrganization(u.ID, o.ID)
 		if err != nil {
 			return err
 		}
-		u, err = d.findUserByID(tx, u.ID)
+		u, err = tx.FindUserByID(u.ID)
 		return err
 	})
 	if err != nil {
@@ -133,12 +127,8 @@ func (d DB) InviteUser(email, orgExternalID string) (*users.User, bool, error) {
 
 // FindUserByID finds the user by id
 func (d DB) FindUserByID(id string) (*users.User, error) {
-	return d.findUserByID(d.DB, id)
-}
-
-func (d DB) findUserByID(db squirrel.BaseRunner, id string) (*users.User, error) {
 	user, err := d.scanUser(
-		d.usersQuery().RunWith(db).Where(squirrel.Eq{"users.id": id}).QueryRow(),
+		d.usersQuery().Where(squirrel.Eq{"users.id": id}).QueryRow(),
 	)
 	if err == sql.ErrNoRows {
 		err = users.ErrNotFound
@@ -151,12 +141,8 @@ func (d DB) findUserByID(db squirrel.BaseRunner, id string) (*users.User, error)
 
 // FindUserByEmail finds the user by email
 func (d DB) FindUserByEmail(email string) (*users.User, error) {
-	return d.findUserByEmail(d.DB, email)
-}
-
-func (d DB) findUserByEmail(db squirrel.BaseRunner, email string) (*users.User, error) {
 	user, err := d.scanUser(
-		d.usersQuery().RunWith(db).Where("lower(users.email) = lower($1)", email).QueryRow(),
+		d.usersQuery().Where("lower(users.email) = lower($1)", email).QueryRow(),
 	)
 	if err == sql.ErrNoRows {
 		err = users.ErrNotFound
@@ -169,13 +155,8 @@ func (d DB) findUserByEmail(db squirrel.BaseRunner, email string) (*users.User, 
 
 // FindUserByLogin finds the user by login
 func (d DB) FindUserByLogin(provider, providerID string) (*users.User, error) {
-	return d.findUserByLogin(d.DB, provider, providerID)
-}
-
-func (d DB) findUserByLogin(db squirrel.BaseRunner, provider, providerID string) (*users.User, error) {
 	user, err := d.scanUser(
 		d.usersQuery().
-			RunWith(db).
 			Join("logins on (logins.user_id = users.id)").
 			Where(squirrel.Eq{
 				"logins.provider":    provider,

--- a/users/db/timed.go
+++ b/users/db/timed.go
@@ -252,7 +252,3 @@ func (t timed) Close() error {
 		return t.d.Close()
 	})
 }
-
-func (t timed) Truncate() error {
-	return t.d.(Truncater).Truncate()
-}

--- a/users/db/traced.go
+++ b/users/db/traced.go
@@ -166,8 +166,3 @@ func (t traced) Close() (err error) {
 	defer func() { t.trace("Close", err) }()
 	return t.d.Close()
 }
-
-func (t traced) Truncate() (err error) {
-	defer func() { t.trace("Truncate", err) }()
-	return t.d.(Truncater).Truncate()
-}

